File: src/eigenv.f90
RECURSIVE SUBROUTINE PDEEigenvalues(Lambda,Q,nv)
  USE iso_c_binding
  IMPLICIT NONE
  INTEGER, PARAMETER             :: nVar = 59                           ! The number of variables of the PDE system 
  TYPE, bind(C) :: tEquations
      REAL(8)    :: CCZ4e=0.1, CCZ4ds=0.11, CCZ4GLMc = 0.75, CCZ4GLMd = 0.75 
  END TYPE tEquations
  TYPE(tEquations) :: EQN
  REAL :: Lambda(nVar), nv(3), Q(nVar), tempA, tempB
  INTENT(IN)  :: Q,nv
  INTENT(OUT) :: Lambda 
  ! Local Variables 
  REAL :: lam, mu, irho, VPR(3), cs,c0,uu,alpha
#if defined(CCZ4EINSTEIN)  
    alpha = MAX( 1.0, EXP(Q(17)) )*MAX( 1.0, EXP(Q(55)) )/MIN( SQRT(Q(1)), SQRT(Q(4)), SQRT(Q(6)) )     
#else
    alpha = 1.0 
#endif
    tempA = alpha*MAX(SQRT(2.0), EQN%CCZ4e, EQN%CCZ4ds, EQN%CCZ4GLMc/alpha, EQN%CCZ4GLMd/alpha )
    tempB = DOT_PRODUCT(Q(18:20),nv(:))
    Lambda = 0.0 
    Lambda(1) = -tempA-tempB 
    Lambda(2) =  tempA-tempB
    !Lambda(1) = -alpha*MAX(SQRT(2.0), EQN%CCZ4e, EQN%CCZ4ds, EQN%CCZ4GLMc/alpha, EQN%CCZ4GLMd/alpha ) - DOT_PRODUCT(Q(18:20),nv(:))   ! MAX( SQRT(2.0), EQN%CCZ4e, EQN%CCZ4ds ) + SQRT(SUM(Q(18:20)**2)) 
    !Lambda(2) = +alpha*MAX(SQRT(2.0), EQN%CCZ4e, EQN%CCZ4ds, EQN%CCZ4GLMc/alpha, EQN%CCZ4GLMd/alpha ) - DOT_PRODUCT(Q(18:20),nv(:))   ! MAX( SQRT(2.0), EQN%CCZ4e, EQN%CCZ4ds ) + SQRT(SUM(Q(18:20)**2)) 
    !
END SUBROUTINE PDEEigenvalues

program main
  implicit none
  REAL :: Q(59), nv(3), Lambda(59),dummy(3)
  integer::i, seed_size
  integer,allocatable :: seed(:)
  call random_seed() ! initialize with system generated seed
  call random_seed(size=seed_size) ! find out size of seed
  allocate(seed(seed_size))
  call random_seed(get=seed) ! get system generated seed
  seed=314159261
  call random_seed(put=seed) ! set current seed
  call random_seed(get=seed) ! get current seed
  deallocate(seed)           ! safe


  call random_number(Q)

  nv = (/1,0,0/)
  dummy=(/2,0,0/)
  !print*,Q

  ! The initial data for Q is chosen such that we do not divide by 0 in the inverse determinant

  !do i=1,1000000000
  call PDEEigenvalues(Lambda, Q, nv)
  !enddo

  print*,Lambda
  print*,nv
  print*, dot_product(nv, dummy)
  !do i=1,59
    !print*,i,Lambda(i)
  !enddo
  !call PDENCP(BgradQ, Q, gradQin)
  !print*, BgradQ
end program main

File: src/eigenv.cxx
#include <cmath>
#include <iostream>

void pdeeigenvalues_(double* lambda, const double* const Q, int normal)
{
  constexpr double sqrtwo = 1.4142135623730951;
  const double qmin = std::min({Q[0],Q[3],Q[5]});
  const double test = Q[16]+Q[54];
#if defined(CCZ4EINSTEIN)  
    //const double alpha = std::fmax( 1.0, std::exp(Q[16]) ) * std::fmax(1.0, std::exp(Q[54])) / std::sqrt(qmin);     
    double alpha =std::exp(fmax( 0, Q[16]+ Q[54]))/ std::sqrt(qmin);
    //double alpha =  1. / std::sqrt(qmin);
    //if (test>0) alpha *= std::exp(test);
#else
    const double alpha = 1.0;
#endif
    const double CCZ4e=0.1, CCZ4ds=0.11, CCZ4GLMc = 0.75, CCZ4GLMd = 0.75;
    const double tempA = alpha * std::max({sqrtwo, CCZ4e, CCZ4ds, CCZ4GLMc/alpha, CCZ4GLMd/alpha});
    const double tempB = Q[17+normal];//DOT_PRODUCT(Q(18:20),nv(:))
    lambda[1] = -tempA-tempB; 
    lambda[2] =  tempA-tempB;
}

int main()
{
  double lambda[3] = {0,0,0};
  double Q[59] = {0.144127786,0.603628755,0.273857057,0.476581693,0.392365575,0.673205435,0.361241937,1.21842027E-02,0.368709803,0.219898820,0.686351418,3.99175286E-02,0.290666640,0.234226882,0.509774327,0.604918957,0.916497707,0.992717385,0.445318341,0.647655010,0.107183695,6.34464025E-02,0.788047969,0.525901496,0.229169130,0.867747128,9.73913670E-02,0.746887863,5.47619462E-02,0.830878258,0.493397176,0.887370348,0.567643166,0.277938724,0.891554296,0.273388088,0.411233306,0.577602625,0.853790164,0.622290909,0.996963739,0.783919871,0.103851914,0.829477191,1.94624662E-02,0.860917091,0.317981422,0.244459093,6.99177980E-02,0.904407263,0.450430274,0.204707682,0.810942411,0.191997290,0.257384479,0.758928418,0.499860823,0.198119998,9.39652920E-02};

  for (int i=0;i<1000000000;i++)
    pdeeigenvalues_(lambda,Q, 0);
  std::cout << lambda[0] << " " << lambda[1] << " "  << lambda[2] << "\n";

  return 0;
}
